docs/decisions/0021-json-serializable-custom-types.md

```diff
@@ -0,0 +1,72 @@
+---
+status: proposed
+contact: dehoward
+date: 2023-11-06
+deciders: alliscode, markwallace-microsoft
+consulted:
+informed:
+---
+
+# JSON Serializable Custom Types
+
+## Context and Problem Statement
+
+This ADR aims to simplify the usage of custom types by allowing developers to use any type that can be serialized using `System.Text.Json`.
+Standardizing on JSON, rather than allowing the developer choose the serialization format, will help enable the use of `JsonSchema` for planner functional manuals.
+
+Today, use of custom types within Semantic Kernel requires developers to implement a custom `TypeConverter` to convert to/from the string representation of the type. This is demonstrated in [Example60_AdvancedNativeFunctions](https://github.com/microsoft/semantic-kernel/blob/main/dotnet/samples/KernelSyntaxExamples/Example60_AdvancedNativeFunctions.cs#L202C44-L202C44) as seen below:
+
+```csharp
+    [TypeConverter(typeof(MyCustomTypeConverter))]
+    private sealed class MyCustomType
+    {
+        public int Number { get; set; }
+
+        public string? Text { get; set; }
+    }
+
+    private sealed class MyCustomTypeConverter : TypeConverter
+    {
+        public override bool CanConvertFrom(ITypeDescriptorContext? context, Type sourceType) => true;
+
+        public override object? ConvertFrom(ITypeDescriptorContext? context, CultureInfo? culture, object value)
+        {
+            return JsonSerializer.Deserialize<MyCustomType>((string)value);
+        }
+
+        public override object? ConvertTo(ITypeDescriptorContext? context, CultureInfo? culture, object? value, Type destinationType)
+        {
+            return JsonSerializer.Serialize(value);
+        }
+    }
+```
+
+## Considered Options
+
+**1. Use native serialization methods**
```

Comment:
In scope of this work, it would be nice to describe in what part of SK the serialization/deserialization logic will be performed. There are a couple of options:
1. As current approach with TypeConverter, we can do serialization/deserialization in NativeFunction. This solution should temporarily resolve the problem, but in NativeFunction and other places in SK we will still be coupled to the string.
2. Do serialization only in place where it's actually needed - in SemanticFunction, to prepare a request and call AI. In this case, we will avoid multiple serialization/deserialization actions in data processing, and objects will be passed between native functions as-is. This will require more changes and design, because there are other open questions, how to represent object as string in meaningful way, how to convert it back to complex type (TypeChat?) etc.

So, it would be great to have more information in this ADR about high-level implementation. Please let me know if that makes sense. Thank you!

Reply:
Tagging @SergeyMenshykh to verify if this part of work doesn't overlap with current refactoring process.

Reply:
@dmytrostruk the idea proposed was going to lean more toward number 1, by to updating `NativeFunction` like you mentioned and likely making updates to the other places as needed to allow the object to be passed as-is. 
number 2 is interesting tho. for your open questions, isn't this what we would use the JSON serialization for? I'm not sure I understand what you mean there.

Reply:
> isn't this what we would use the JSON serialization for? I'm not sure I understand what you mean there.

Let's take the case, when we have following functions in pipeline:
1. `NativeFunction1` - it returns `ComplexType1`.
2. `SemanticFunction1` - it should accept `ComplexType1`, pass it to AI, and return `string` or `ComplexType2` (depends on how we will implement handling of objects instead of strings).
3. `NativeFunction2` - it accepts `ComplexType2` and returns some result.

In case of JSON serialization, in `SemanticFunction1` you will take `ComplexType1`, serialize it to JSON and pass it to AI. Then, you will get some response, and in order to pass that response to `NativeFunction2`, you will need to convert it to `ComplexType2`.

The open question is - how to do that properly? If we use JSON deserialization for that, the response from `SemanticFunction1` should be always in JSON format. We will have to force model to respond in JSON only to achieve that. Otherwise, if you will do `JsonSerializer.Deserialize<ComplexType2>("response_from_AI")`, where `response_from_AI` is plain string, you will receive `System.Text.Json.JsonException: ''response_from_AI' is an invalid JSON literal.`, and `JsonConverter` won't handle that. While with `TypeConverter` it's still possible to do it, because `converter.ConvertFromString("response_from_AI")` will pass this value to your converter, and you will be able to map it to your complex type correctly.

The other way would be to replace `ComplexType2` parameter in `NativeFunction2` with `string` and delegate data processing to the user, but in this case, we won't support complex types natively, it's the same `string` processing, so it would be nice to avoid that.

Reply:
after some talks following this comment, it was decided that we're ok with constraining the model to always return JSON responses given that we will know how to deserialize the response. I've added cases in the ADR where serialization/deserialization will be necessary, specifically when we're going from a native fn response to a semantic fn input vice versa.

the updated proposed approach that's stemmed from the discussions is: **Use TypeConverter for primitive types and native serialization methods for custom types**